lib/utilities/retrieveData.js

```diff
@@ -1,31 +1,73 @@
-export default function retrieveData(locations, dataRetrievalFn,
-    response, logger, errorHandlerFn) {
-    if (locations.length === 0) {
-        return response.end();
-    }
-    if (errorHandlerFn === undefined) {
-        // eslint-disable-next-line
-        errorHandlerFn = () => { response.connection.destroy(); };
-    }
-    const current = locations.shift();
-    return dataRetrievalFn(current, logger,
-        (err, readable) => {
+const { eachSeries } = require('async');
+const responseErr = new Error();
+responseErr.code = 'ResponseError';
+responseErr.message = 'response errored out before readable finished';
+
+export default function retrieveData(locations, retrieveDataFn, response, log) {
+    let responseDestroyed = false;
+    const _destroyResponse = () => {
+        // destroys the socket if available
+        response.destroy();
+        responseDestroyed = true;
+    };
+    response.on('close', () => {
+        log.debug('received close event before response end');
+        _destroyResponse();
+    });
+
+    eachSeries(locations,
+        (current, next) => retrieveDataFn(current, log, (err, readable) => {
+            let cbCalled = false;
+            const _next = err => {
+                // Avoid multiple callbacks since it's possible that response's
+                // close event and the readable's end event are emitted at
+                //  the same time.
+                if (!cbCalled) {
+                    next(err);
+                    cbCalled = true;
```

Comment:
just to be safe, set this to true before calling next(err)?

Reply:
Sure.